base/applications/rapps/cabinet.cpp

```diff
@@ -0,0 +1,351 @@
+                           PFNREAD, 
+                           PFNWRITE,
+                           PFNCLOSE, 
+                           PFNSEEK, 
+                           int, 
+                           PERF);
+
+typedef BOOL(*fnFDICopy)(HFDI,
+                         LPSTR,
+                         LPSTR,
+                         INT,
+                         PFNFDINOTIFY,
+                         PFNFDIDECRYPT,
+                         void FAR *pvUser);
+
+typedef BOOL(*fnFDIDestroy)(HFDI);
+
+/* 
+ * Extraction function 
+ * TODO: require only a full path to the cab as an argument
+ */
+BOOL ExtractFilesFromCab(const ATL::CStringW& szCabName, 
+                         const ATL::CStringW& szCabDir, 
+                         const ATL::CStringW& szOutputDir)
+{
+    HINSTANCE hCabinetDll;
+    HFDI ExtractHandler;
+    ERF ExtractErrors;
+    ATL::CStringA szCabNameUTF8, szCabDirUTF8, szOutputDirUTF8;
+    fnFDICreate pfnFDICreate;
+    fnFDICopy pfnFDICopy;
+    fnFDIDestroy pfnFDIDestroy;
+    BOOL bResult;
+
+    // Load cabinet.dll and extract needed functions 
+    hCabinetDll = LoadLibraryW(L"cabinet.dll");
+
+    if (!hCabinetDll)
+    {
+        return FALSE;
+    }
+
+    pfnFDICreate = (fnFDICreate) GetProcAddress(hCabinetDll, "FDICreate");
+    pfnFDICopy = (fnFDICopy) GetProcAddress(hCabinetDll, "FDICopy");
+    pfnFDIDestroy = (fnFDIDestroy) GetProcAddress(hCabinetDll, "FDIDestroy");
+
+    if (!pfnFDICreate || !pfnFDICopy || !pfnFDIDestroy)
+    {
+        FreeLibrary(hCabinetDll);
+        return FALSE;
+    }
+
+    // Create FDI context
+    ExtractHandler = pfnFDICreate(fnMemAlloc,
+                                  fnMemFree,
+                                  fnFileOpen,
+                                  fnFileRead,
+                                  fnFileWrite,
+                                  fnFileClose,
+                                  fnFileSeek,
+                                  cpuUNKNOWN,
+                                  &ExtractErrors);
+
+    if (!ExtractHandler)
+    {
+        FreeLibrary(hCabinetDll);
+        return FALSE;
+    }
+
+    // Create output dir
+    CreateDirectoryW(szOutputDir, NULL);
+
+    // Convert wide strings to UTF-8
+    bResult = WidePathToUTF8(szCabName, szCabNameUTF8);
+    bResult &= WidePathToUTF8(szCabDir, szCabDirUTF8);
+    bResult &= WidePathToUTF8(szOutputDir, szOutputDirUTF8);
+
+    // Add a slash to cab name as required by the api
+    szCabNameUTF8 = "\\" + szCabNameUTF8;
+
+    // Perform extraction
+    if (bResult)
+    {
+        bResult = pfnFDICopy(ExtractHandler,
+                             (LPSTR) szCabNameUTF8.GetString(),
+                             (LPSTR) szCabDirUTF8.GetString(),
+                             0,
+                             fnNotify,
+                             NULL,
+                             (void FAR *) szOutputDirUTF8.GetString());
+
+        bResult &= pfnFDIDestroy(ExtractHandler);
```

Comment:
I'm not sure it is worth looking at the success value of the FDIDestroy call (since technically stuff have been already extracted, as you wanted).